src/Data/Bifunctor/Lazier.hs

```diff
@@ -0,0 +1,163 @@
+  x <$ Lazier m = Lazier (x <$ lazify m)
+
+instance (Bifunctor p, forall a b. Lazify (p a b)) => Bifunctor (Lazier p) where
+  bimap f g (Lazier x) = Lazier (bimap f g (lazify x))
+  first f (Lazier x) = Lazier (first f (lazify x))
+  second g (Lazier x) = Lazier (second g (lazify x))
+
+instance (Applicative (p a), forall b. Lazify (p a b)) => Applicative (Lazier p a) where
+  pure :: forall b. b -> Lazier p a b
+  pure = coerce (pure @(p a) @b)
+
+  Lazier fs <*> Lazier xs = Lazier (lazify fs <*> lazify xs)
+  liftA2 f (Lazier xs) (Lazier ys) = Lazier (liftA2 f (lazify xs) (lazify ys))
+
+  Lazier xs *> Lazier ys = Lazier (lazify xs *> lazify ys)
+  Lazier xs <* Lazier ys = Lazier (lazify xs <* lazify ys)
+
+instance (Biapplicative p, forall a b. Lazify (p a b)) => Biapplicative (Lazier p) where
+  bipure :: forall a b. a -> b -> Lazier p a b
+  bipure = coerce (bipure @p @a @b)
+
+  Lazier fs <<*>> Lazier xs = Lazier (lazify fs <<*>> lazify xs)
+  biliftA2 f g (Lazier xs) (Lazier ys) = Lazier (biliftA2 f g (lazify xs) (lazify ys))
+
+  Lazier xs *>> Lazier ys = Lazier (lazify xs *>> lazify ys)
+  Lazier xs <<* Lazier ys = Lazier (lazify xs <<* lazify ys)
+
+-- | Product types, newtypes around product types, etc., that can be made lazy.
+-- There is one instance, which works for most lazifiable types that are
+-- instances of `Generic`. In the event that you need an instance for a type
+-- for which that doesn't work, you can write an overlapping one. For example,
+--
+-- @
+-- data Baz = Beep !Void | Boop Int
+-- @
+--
+-- can't be lazified generically, but you can write
+--
+-- @
+-- instance {-\# OVERLAPPING \#-} Lazify Baz where
+--   lazify ~(Boop i) = Boop i
+-- @
+--
+-- This can also be used to write instances for types that aren't instances
+-- of 'Generic'.
+--
+-- @since 6
+class Lazify a where
+  -- Make the value of a product type lazy. For example, for pairs,
+  -- it is (effectively) defined
+  --
+  -- @
+  -- lazify ~(a, b) = (a, b)
+  -- @
+  lazify :: a -> a
+
+instance (Generic a, GLazify a (Rep a)) => Lazify a where
+  lazify = to . glazify @a . from
+
+class GLazify a f where
+  glazify :: f p -> f p
+
+instance GLazify a f => GLazify a (D1 ('MetaData _q _r _s 'False) f) where
+  glazify (M1 m) = M1 (glazify @a m)
+instance GLazify a f => GLazify a (C1 c f) where
+  glazify (M1 m) = M1 (glazify @a m)
+instance GLazify a f => GLazify a (S1 ('MetaSel _p _q _r 'DecidedLazy) f) where
+  glazify (M1 m) = M1 (glazify @a m)
+instance TypeError ('Text "Can't lazify " ':<>: 'ShowType a ':<>: 'Text ":"
+                    ':$$: 'Text "It has a strict field.")
+  => GLazify a (S1 ('MetaSel _p _q _r 'DecidedStrict) f) where
+  glazify _ = error "Unreachable"
+instance TypeError ('Text "Can't lazify " ':<>: 'ShowType a ':<>: 'Text ":"
+                    ':$$: 'Text "It has a strict (unpacked) field.")
+  => GLazify a (S1 ('MetaSel _p _q _r 'DecidedUnpack) f) where
+  glazify _ = error "Unreachable"
```

Comment:
Ooh, this is a mistake... We can lazify these. We can't lazify their `bimap` or `biliftA2` successfully, but that's another story.

Reply:
Er... No, we can't. What was I thinking?

Reply:
Ah, I see what we're supposed to do with those! If there's a strict field, then we should try to lazify it!

Reply:
Actually, I'm not sure. Suppose we have

```haskell
data Boof a b = Boof () a b
instance Bifunctor Boof where
  bimap f g (Boof () a b) = Boof () (f a) (g a)
```

Now `bimap f g . Lazier` will be strict in its argument, because it matches on `()`. Lazifying the argument first won't have any effect; forcing that field forces the whole thing. Now if we define

```haskell
data Boof a b = Boof !() a b
```

and we try to be clever, then the exact same `Bifunctor Boof` definition leads to `bimap f g . Lazier` being lazy! This doesn't seem right, does it?

Reply:
No, I think it's okay to just reject strict fields. If someone wants a lazy instance for something with a strict field, they can write a specialized custom wrapper.